### PR TITLE
feat(empty-state): Progressive enhancements for relayout based on ResizeObserver

### PR DESCRIPTION
**Pull request by tomheller**
### <strong>Pull Request</strong>

Uses the ResizeObserver to actively match the width of the emptyState component and adjust the layout accordingly. 

#### Type of PR

Feature

#### Checklist

- [x] I have read the CONTRIBUTING doc and I follow the PR guidelines
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: components/empty-state/src/empty-state.ts [hunk 1]
```diff
@@ -50,6 +50,8 @@ const ITEMS_HORIZONTAL_BREAKPOINT = 540;
 /** The min-width from which the empty state items are aligned next to each other. */
 const LAYOUT_HORIZONTAL_BREAKPOINT = 760;
 
+declare const window: any;
```
Comment (lukasholzer) [inline components/empty-state/src/empty-state.ts]: NIT: you can declare your own typings till it is part of the `lib.dom` package: microsoft/TypeScript#28502


```typescript
/** Workaround until created: microsoft/TypeScript#28502 */
interface ResizeObserver {
...
}
```

```suggestion
declare const window: { ResizeObserver?: ResizeObserver };
```

File: components/empty-state/src/empty-state.ts [hunk 2]
```diff
@@ -186,7 +190,11 @@ export class DtEmptyState
   constructor(
     private _changeDetectorRef: ChangeDetectorRef,
     private _elementRef: ElementRef<HTMLElement>,
-    private _viewportResizer: DtViewportResizer,
+    /**
+     * @deprecated Remove the viewportResizer from the constructor as it is no longer needed.
+     * @breaking-change Remove the viewportResizer in version 8.0.0
```
Comment (lukasholzer) [inline components/empty-state/src/empty-state.ts]: can we do this with version 7? :D

File: components/empty-state/src/empty-state.ts [hunk 3]
```diff
@@ -197,48 +205,60 @@ export class DtEmptyState
...
+      this._containerSizeObserver = new window.ResizeObserver(entries => {
+        // We need to wrap the call to the layout update into an additional
+        // requestAnimationFrame, because the resize observer would trow a
+        // javascript error when it is not able to process all entries within
+        // a single animation frame.
+        // From the specification:
+        // > This error means that ResizeObserver was not able to deliver all
+        // > observations within a single animation frame. It is benign (your
+        // > site will not break). - Aleksandar Totic
+        // https://stackoverflow.com/a/50387233
```
Comment (lukasholzer) [inline components/empty-state/src/empty-state.ts]: Nice explanation!

File: components/empty-state/src/empty-state.ts [hunk 4]
```diff
@@ -197,48 +205,60 @@ export class DtEmptyState
...
+        window.requestAnimationFrame(() => {
+          this._updateLayout(entries[0].contentRect.width);
```
Comment (lukasholzer) [inline components/empty-state/src/empty-state.ts]: Is it possible that entries is empty or has no index 0?

Comment (tomheller) [reply components/empty-state/src/empty-state.ts]: Not really, that would mean that the elementRef of the host element is empty. But I can add a check there, just to be sure.

File: components/empty-state/src/empty-state.ts (hunk 2 above)
Comment (tomheller) [reply components/empty-state/src/empty-state.ts]: I believe that at some point we have discussed this in the past and decided that we will announce breaking-changes 2 major versions in advance.

File: libs/barista-components/empty-state/src/empty-state.ts [hunk 5]
```diff
@@ -197,61 +202,52 @@ export class DtEmptyState
   }
 
   ngAfterViewInit(): void {
-    this._viewportResizer
-      .change()
-      .pipe(startWith(null), takeUntil(this._destroy$))
-      .subscribe(() => {
-        this._updateLayout();
+    // Check if the broser supports the resizeObserver.
```
Comment (lara-aigmueller) [inline libs/barista-components/empty-state/src/empty-state.ts]: nit: browser

**Comment by tomheller** [comment]
This pull request removes the Base class for the dtEmptyState, which is considered a breaking change. Please add it back in, so we don't have a breaking version in there.

File: components/empty-state/src/empty-state.ts (hunk 1 above)
Comment (tomheller) [reply components/empty-state/src/empty-state.ts]: Nope, doesn't work. Tried to type it with 4 typescript specialists at my side... did not work. At some point of our nested tsconfig structure, either the types, files, typeroots or includes are being overwritten... 
this is sad 😢

**Comment by tomheller** [comment]
Put back the deprecated class and functions... Will be removable in version 8

**Comment by lukasholzer** [comment]
formatting and unit tests are failing

**Comment by tomheller** [comment]
fixed formating and unit test. Once more for good measure @lukasholzer @lara-aigmueller

**Pull request merged**